### PR TITLE
Create dynamic micro-factions for dynamic NPCs

### PR DESCRIPTION
**Pull request by ghost**
#### Summary
```SUMMARY: Features "Create dynamic micro-factions for dynamic NPCs"```

#### Purpose of change
There is a longer-term plan to make dynamic NPCs behave realistically and form and build up factions, and go and loot and make bases etc.
First step in this, is to allocate each dynamically spawned NPC their own micro-faction/ lone-wolf faction, unique to them.
( in the future, they will join up with others, and merge factions etc )

#### Describe the solution
Dynamic NPCs and Debug spawned NPCs now get a new faction created for them, based on their name. This does nothing really at the moment, but is a building block for further stuff in this area.
When they die, if it is still a "lone_wolf_faction" i.e they are the only member, the faction is then deleted, to stop the faction list bloating.

#### Describe alternatives you've considered
N/A

#### Additional context
I discovered something odd.
The previous bugs about dynamic NPCs going hostile immediately, and some members of the refugee center was due to an assumption that all dynamic NPCs belonged to the "wasteland_scavenger" faction, which then communicated wrongdoing across all its future members.
which is why I submitted #33738 
Upon investigation however, I discovered that dynamic NPCs do not get any faction allocated at all, not wasteland_scavengers, not the new "no_faction" not anything.
Which is very odd, as that was assumed to be the cause of those hostility bugs.

I'm not sure if this is a new thing or not, but its kinda tangential to this PR, as now they do get a faction assigned to them, but it is very puzzling.

### CODE REVIEW

File: src/faction.cpp [hunk 1]
```diff
@@ -316,6 +322,21 @@ void faction_manager::clear()
...
 
+void faction_manager::remove_faction( const faction_id &id )
+{
+    if( id.str().empty() || id == faction_id( "no_faction" ) ) {
+        return;
+    }
+    int index = 0;
+    for( faction &elem : factions ) {
+        if( id == elem.id ) {
+            factions.erase( factions.begin() + index );
```
[the remark this answers is not in the document]
Comment (BevapDin) [reply src/faction.cpp]: > does erasing from a map avoid pointer invalidation?

Yes, removing elements from a `std::map` does not affect the remaining elements.

[the remark this answers is not in the document]
Comment (davidpwbrown) [reply src/faction.cpp]: Yep, took some adjusting, but worked that in.

**Comment by BevapDin** [comment]
@davidpwbrown I just recognized something relevant for this:

When you leave a follower via the talking (telling them "I'm going on my own for a while"), the game calls `talk_function::leave`. This in turn calls `npc::clear_fac`, which in turn sets the faction to null and the faction id to an empty string.

This should probably revert the faction to a dynamic micro-faction so the NPC still has a valid faction.

**Comment by davidpwbrown** [comment]
> @davidpwbrown I just recognized something relevant for this:
> 
> When you leave a follower via the talking (telling them "I'm going on my own for a while"), the game calls `talk_function::leave`. This in turn calls `npc::clear_fac`, which in turn sets the faction to null and the faction id to an empty string.
> 
> This should probably revert the faction to a dynamic micro-faction so the NPC still has a valid faction.

Right, Thanks ill take a look at that.

**Pull request merged**